velox/exec/Window.cpp

```diff
@@ -376,15 +434,20 @@ void Window::callApplyForPartitionRows(
     rawPeerEnds[j] = peerEndRow_ - 1 - firstPartitionRow;
 
     for (auto w = 0; w < numFuncs; w++) {
-      auto frameEndPoints = findFrameEndPoints(
-          w,
-          firstPartitionRow,
-          lastPartitionRow,
-          peerStartRow_,
-          peerEndRow_ - 1,
-          i);
-      rawFrameStartBuffers[w][j] = frameEndPoints.first - firstPartitionRow;
-      rawFrameEndBuffers[w][j] = frameEndPoints.second - firstPartitionRow;
+      rawFrameStartBuffers[w][j] = windowFrames_[w].startFrameFunction(
```

Comment:
The way the code is structured we are making a function call per row and per window function. That's a lot of function calls. A more optimal code structure would be to loop over window functions and make a single function call to process all rows.

startFrameFunction and endFrameFunction need to be modified to take a list of rows and populate a list of frameStart/Ends in one call.

Functions like findRowUnboundedPreceding can then use std::memset to fill in multiple vales at once.

Reply:
Thanks for the suggestion, the function call overhead can be reduced for unbounded preceding and unbounded following frame bounds using this approach. But since CURRENT ROW frame bound depends on the peerStartRow and peerEndRow which are computed here and also because k PRECEDING and k FOLLOWING frame bound computation (#2928) uses the index 'j' from this loop, I am not sure how this function call can be moved outside the loop for these cases. 
Can I add this as an optimization for kUnbounded Preceding/Following frames only?

Reply:
I assume you can pass rawPeerStarts and rawPeerEnds to the function, no?

Reply:
Yeah, will move the frame computation outside this loop and modify the frame bound functions accordingly, thanks!

Reply:
Though I feel that if we are going to do a single function call to process all rows, then maybe we don't need to do the function pointer approach and have different functions based on the frame type. The parameters for the functions will be quite different based on ROW, k PRECEDING, k FOLLOWING or if a column is used to populate k.

Masha,
  wdyt ?

Reply:
I don't have a strong preference. Either way works.

Reply:
Thanks for the suggestion, removed the use of function pointers.